File: 이건희/level2/숫자 카드 나누기.py
def solution(arrayA, arrayB):
    numA = calculate(commonDivisor(arrayA), arrayB)
    numB = calculate(commonDivisor(arrayB), arrayA)
    return max(numA + numB)

def divisor(number):
    result = []
    for i in range(1, int(number**(1/2))+1):
        if number%i==0:
            result.append(i)
            if i < number//i:
                result.append(number//i)
                
    return sorted(result) 

def commonDivisor(array):
    numList = [0]
    for i in divisor(min(array))[1:]:
        for j in array:
            if j % i != 0:
                break
        else:
            numList.append(i)
            
    return numList

def calculate(numList, array):
    result = [0]
    if len(numList) == 1:
        return [0]
    for i in numList[1:]:
        for j in array:
            if j % i == 0:
                break
        else:
            result.append(i)
            
    return result

File: 이훈일/Programmers/Level2/숫자 카드 나누기.java
import java.util.*;

class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        Arrays.sort(arrayA);
        Arrays.sort(arrayB);
        
        int numA = arrayA[0];
        List<Integer> numsA = new ArrayList<>();
        for(int i = numA; i >= 1; i--) {
            if(numA % i == 0) {
                numsA.add(i);
            }
        }
        
        int numB = arrayB[0];
        List<Integer> numsB = new ArrayList<>();
        for(int i = numB; i >= 1; i--) {
            if(numB % i == 0) {
                numsB.add(i);
            }
        }
        
        int answerA = 0;
        for(int i = 0; i < numsA.size(); i++) {
            boolean isPossible = true;
            int comp = numsA.get(i);
            
            for(int j = 0; j < arrayA.length; j++) {    
                if(arrayA[j] % comp != 0) {
                    isPossible = false;
                    break;
                }
                
                if(arrayB[j] % comp == 0) {
                    isPossible = false;
                    break;
                }
            }
            
            if(isPossible) {
                answerA = comp;
                break;
            }
        }
        
        int answerB = 0;
        for(int i = 0; i < numsB.size(); i++) {
            boolean isPossible = true;
            int comp = numsB.get(i);
            
            for(int j = 0; j < arrayB.length; j++) { 
                if(arrayB[j] % comp != 0) {
                    isPossible = false;
                    break;
                }
                
                if(arrayA[j] % comp == 0) {
                    isPossible = false;
                    break;
                }
            }
            
            if(isPossible) {
                answerB = comp;
                break;
            }
        }
        
        return Math.max(answerA, answerB);
    }
}
